### DISCUSSION TITLE
Convert binary representation to armored and back

### DISCUSSION THREAD

**Discussion by AndreasGassmann** [openpgpjs/openpgpjs#1397] [Q&A]
- OpenPGP.js version: 5.0.0-5
- Affected platform (Browser or Node.js version): Browser, Angular application

## Setup

I am encrypting a text with the following code:

```ts
    const message = await createMessage({ text })
    const encrypted = await encrypt({
      message, // input as Message object
      passwords: [password], // multiple passwords possible
      config: { preferredCompressionAlgorithm: enums.compression.zlib },
      format: 'binary'
    })
    console.log('encrypted', encrypted)
```

This gives me a UInt8Array output, as expected.

The test values I used are:

- Text: `test`
- Password: `secret`
- Armored output (converted with the method below):
```
-----BEGIN PGP MESSAGE-----

wy4ECQMIXgEtDG2fDJTgb4lQyosDy1mPyeSuIzKIytVpkV9l72+L7NzXp5Xe
CQhd0kYBNqqsyIhrOnaCqEr5RWCI+MM2A8LGVuI0AUbSmPkw/kxO05lDT2K7
Pps2aRHIa+51iQ8w+J2y0C1nPsqHNb+8ThyGy3fM
=Bhh6
-----END PGP MESSAGE-----
```

## Problem

I would now like to convert the output above to an armored representation and back.

I tried to following code:

```ts
  async convertBinaryToArmored(binaryMessage: Uint8Array): Promise<string> {
    const message = await readMessage({
      binaryMessage
    })

    console.log(message.armor())
    console.log(armor(enums.armor.message, message.write(), null, null, null))
    console.log(armor(enums.armor.message, binaryMessage, null, null, null))
    
    return armor(enums.armor.message, binaryMessage, null, null, null)
   }
```

It seems that the first 2 logs don't work as expected, but the last one does. The output in the console in the browser is the following:

![Screenshot 2021-08-08 at 00 54 37](https://user-images.githubusercontent.com/680814/128615787-de0b4af0-3413-4b72-91c1-b094b2e14439.png)

it looks like there is some kind of issue related to promises (note the "ZoneAwarePromise", this is something that Angular injects). The last line seems to be working fine. Is it possible that the `message.write()` method, which is also used internally in the `message.armor()` call, is using a promise somewhere that is not properly awaited? Or am I doing something wrong?

## Armored to Binary

And just for completeness, this is how I go from armored to binary, in case anyone needs this in the future. (This seems to work).

```ts
  async convertArmoredToBinary(message: string): Promise<Uint8Array> {
    const data = await unarmor(message)
    return readToEnd(data.data)
  }
```

**Comment by larabr** [comment]
Hi @AndreasGassmann, `Message.armor()` and `.write()` return a `ReadableStream`, so you need to read that out to access the data. Does this answer your question?

And again for completeness: to go from armored to binary, you can also just do `openpgp.readMessage({ armoredMessage })` and then `.write()`.

**Reply by AndreasGassmann** [reply]
Thanks for the reply. This does work, but then that means the typescript typings are wrong, because the output should be a string:

https://github.com/openpgpjs/openpgpjs/blob/master/openpgp.d.ts#L256

**Reply by larabr** [reply]
Thanks for spotting that, the typings don't match with the JSDocs indeed. Feel free to open a PR :)

**Reply by AndreasGassmann** [reply]
I'm still in the middle of integrating openpgp into our app and I also found some other issues with the typings (eg. in the signing example in the readme). Once I'm done I'll create a PR with all the issues I found. 👍
